File: internal/shadow/core/dispatch.go
package core

import "fmt"

// ShadowContext is the per-consumer context threaded through a shadow run.
type ShadowContext struct {
	WorkflowsRepo string
	WorkflowsRef  string
	ConsumerRepo  string
	ConsumerRef   string
	WorkflowsPR   int
	Branch        string
}

// BuildDispatchInputs maps the context to the receiver workflow_dispatch `inputs` (all strings).
func BuildDispatchInputs(ctx ShadowContext) map[string]string {
	return map[string]string{
		"workflows_repo": ctx.WorkflowsRepo,
		"workflows_ref":  ctx.WorkflowsRef,
		"consumer_repo":  ctx.ConsumerRepo,
		"consumer_ref":   ctx.ConsumerRef,
		"workflows_pr":   fmt.Sprintf("%d", ctx.WorkflowsPR),
		"branch":         ctx.Branch,
	}
}

// RunState is a workflow run's coarse state, driving quiet polling.
type RunState string

const (
	RunPending RunState = "pending"
	RunSuccess RunState = "success"
	RunFailure RunState = "failure"
)

// ClassifyRunState classifies a run from its status/conclusion. An empty conclusion means "null"
// (not yet concluded or no result) — treated as failure once status is "completed".
func ClassifyRunState(status, conclusion string) RunState {
	if status != "completed" {
		return RunPending
	}
	if conclusion == "success" {
		return RunSuccess
	}
	return RunFailure
}

// ExtractRunID reads the numeric workflow_run_id from a `return_run_details:true` dispatch response
// (REST shape `{workflow_run_id, run_url, html_url}`, JSON-decoded into a map).
func ExtractRunID(response any) (int, error) {
	if m, ok := response.(map[string]any); ok {
		if id, ok := m["workflow_run_id"].(float64); ok {
			return int(id), nil
		}
	}
	return 0, fmt.Errorf("workflow_dispatch response has no numeric workflow_run_id: %v", response)
}

File: shadow/src/core/dispatch.ts
export interface ShadowContext {
  workflowsRepo: string;
  workflowsRef: string;
  consumerRepo: string;
  consumerRef: string;
  workflowsPr: number;
  branch: string;
}

/** Map the shadow context to the receiver workflow_dispatch `inputs` (all values are strings). */
export function buildDispatchInputs(ctx: ShadowContext): Record<string, string> {
  return {
    workflows_repo: ctx.workflowsRepo,
    workflows_ref: ctx.workflowsRef,
    consumer_repo: ctx.consumerRepo,
    consumer_ref: ctx.consumerRef,
    workflows_pr: String(ctx.workflowsPr),
    branch: ctx.branch,
  };
}

/**
 * Read the run id from a `workflow_dispatch` response created with `return_run_details: true`.
 * REST shape: `{ workflow_run_id, run_url, html_url }`.
 */
export type RunState = 'pending' | 'success' | 'failure';

/** Classify a workflow run from its `status`/`conclusion`. Pure: drives quiet polling (log only on
 * change, return on success, throw on failure) instead of the noisy `gh run watch` redraw. */
export function classifyRunState(status: string, conclusion: string | null): RunState {
  if (status !== 'completed') return 'pending';
  return conclusion === 'success' ? 'success' : 'failure';
}

export function extractRunId(response: unknown): number {
  const id = (response as { workflow_run_id?: unknown } | null)?.workflow_run_id;
  if (typeof id !== 'number' || !Number.isFinite(id)) {
    throw new Error(`workflow_dispatch response has no numeric workflow_run_id: ${JSON.stringify(response)}`);
  }
  return id;
}
